File: top_interview_questions_easy/sqrt.cpp
class Solution {
public:
    int mySqrt(int x) {
      int low = 0, high = x, mid;
      if(x < 2) return x; // mid가 0이 되면 안되므로 
      while(low < high) {
        // low와 high의 중간값을 구한다. 
        mid = (low + high)/2;
        // x/mid가 mid보다 크면 low의 값을 증가시킨다. 
        if(x/mid >= mid) {
          low = mid + 1;
          // 그렇지 않은 경우엔 high를 mid값으로 낮춘다. 
        }else {
          high = mid;
        }
        // 이렇게 low가 high를 추월하지 않는 동안 반복한다. 
      }
      // 그렇게 찾은 가장 높은 값의 -1 (0부터 시작했으므로)이 제곱근이다. 
      return high - 1;


    }
};

File: top_interview_questions_easy/sqrt.py
"""
Binary Search를 이용한 문제
left, right 를 두고 left가 right를 초월할 때까지 while loop를 돌린다.
중간값 mid를 정한 다음 mid * mid 값이 x가 될 때까지 left와 right를 조정한다.

"""

class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return x

        left = 1
        right = x

        while left <= right:
            mid = (left + right) // 2
            calc = mid * mid
            if calc == x:
                return mid
            elif calc < x:
                left = mid + 1
            else:
                right = mid - 1
        return right
